src/Evented.ts

```diff
@@ -117,36 +44,48 @@ export function isGlobMatch(globString: string | symbol, targetString: string |
 	}
 }
 
+export type EventedCallback<T = EventType, E extends EventObject<T> = EventObject<T>> = {
+	/**
+	 * A callback that takes an `event` argument
+	 *
+	 * @param event The event object
+	 */
+
+	(event: E): boolean | void;
+};
+
+/**
+ * A type which is either a targeted event listener or an array of listeners
+ * @template T The type of target for the events
+ * @template E The event type for the events
+ */
+export type EventedCallbackOrArray<T = EventType, E extends EventObject<T> = EventObject<T>> = EventedCallback<T, E> | EventedCallback<T, E>[];
+
 /**
  * Event Class
  */
-export class Evented extends Destroyable implements Evented {
+export class Evented<M extends {} = {}, T = EventType, O extends EventObject<T> = EventObject<T>> extends Destroyable {
 
-	/**
-	 * map of listeners keyed by event type
-	 */
-	protected listenersMap: Map<string, EventedCallback<EventObject>> = new Map<string, EventedCallback<EventObject>>();
+	// The following member is purely so TypeScript remembers the type of `M` when extending so
+	// that the utilities in `on.ts` will work
+	// tslint:disable-next-line
+	protected __typeMap__?: M;
 
 	/**
-	 * @constructor
-	 * @param options The constructor argurments
+	 * map of listeners keyed by event type
 	 */
-	constructor(options: EventedOptions = {}) {
-		super();
-		const { listeners } = options;
-		if (listeners) {
-			this.own(this.on(listeners));
-		}
-	}
+	protected listenersMap: Map<T, EventedCallback<T, O>> = new Map();
 
 	/**
 	 * Emits the event objet for the specified type
 	 *
 	 * @param event the event to emit
 	 */
-	emit<E extends EventObject>(event: E): void {
+	emit<K extends keyof M>(event: M[K]): void;
+	emit(event: O): void;
+	emit(event: any): void {
 		this.listenersMap.forEach((method, type) => {
-			if (isGlobMatch(type, event.type)) {
+			if (isGlobMatch(<any> type, event.type)) {
```

Comment:
Why do we need `<any>` here?

Reply:
Yes, lets try to comment whenever we have to cast as `any` (as well as prefer the `type as any` instead of the arrow brackets.

Reply:
Because `type` is a generic that can be anything and `isGlobMatch()` requires `string | symbol`

Reply:
> let's try to comment whenever we have to cast as `any`

As in, add a comment to code for future generations, please.